src/main.rs

```diff
@@ -59,7 +69,42 @@ impl StateT for SquareOperatorState {
     fn commit(&mut self, _timestamp: &Timestamp) {}
 }
 
+struct DeadlineCtx {
+    val: usize,
+}
+
+impl DeadlineCtx {
+    pub fn new() -> Self {
+        Self { val: 0 }
+    }
+
+    pub fn increment(&mut self) {
+        self.val += 1;
+    }
+
+    pub fn deadline(&mut self) -> Duration {
+        self.increment();
+        Duration::new(5, 0)
+    }
+}
+
 impl OneInOneOut<SquareOperatorState, usize, usize> for SquareOperator {
+    fn setup(&mut self, ctx: &mut SetupContext<SquareOperatorState>) {
+        let deadline_ctx = Arc::clone(&self.deadline_ctx);
+        ctx.add_deadline(TimestampDeadline::new(
+            move |_s: &SquareOperatorState, _t: &Timestamp| -> Duration {
+                deadline_ctx.lock().unwrap().deadline()
```

Comment:
I'm not a fan of the user manually locking state. Is there anyway we can "hide" the context or pass it to the system via the SetupContext?

Reply:
The user doesn't manually lock the operator state, but this example shows that users can install deadline definition functions and still have state for them separately, if need be. 

The other way of doing it is to force everyone to define a structure that implements a `DeadlineContext` trait (just like we used to have before).

Reply:
Yes, I'm also not a fan of locking the deadline contexts. What is the argument against implementing the DeadlineContext trait?